README.md
simple-sidebar
==============

A simple jQuery sidebar plugin

With this simple plug-in you can choose to use the right-sidebar or the left-sidebar and switch from one to the other in just one type! [see options](#options)

##DEMO:
* [DEMO](http://dcdeiv.github.io/simple-sidebar);
* [Right-Sidebar DEMO](http://dcdeiv.github.io/simple-sidebar/demo-right/index.html);
* [Left-Sidebar DEMO](http://dcdeiv.github.io/simple-sidebar/demo-left/index.html);
* [AJAX](http://dcdeiv.github.io/simple-sidebar/demo-ajax/index.html);
* [Material-Designed](http://materialdesigned.tumblr.com/)

##DOWNLOAD
* **Git**: `git clone https://github.com/dcdeiv/simple-sidebar`;
* **Bower**: `bower install simple-sidebar`.

##CHANGELOG
I strongly recommend to everyone to read the changelog [here](CHANGELOG.md) before updating the plug-in.

##EXAMPLE OF USAGE:
**These options (except mask and mask.style) are all required**

	$( '#sidebar' ).simpleSidebar({
		settings: {
			opener: '.button',
			wrapper: '#wrapper',
		},
		sidebar: {
			align: 'left'
			width: 250,
			closingLinks: 'a',
			style: {
				zIndex: 100
			}
		},
		mask: {
			style: {
				backgroundColor: 'grey',
				opacity: 0.9,
				filter: 'Alpha(opacity=90)'
			}
		}
	});

#OPTIONS

	//Select the $sidebar. Ex: '.sidebar'
	$( 'sidebar-selector' ).simpleSidebar({
		//Global plug-in settings
		//*REQUIRED*
		settings: {
			//*REQUIRED* OPENER is the selector that fire the animation. It can be an icon a button, an input, an image, a div or a span. Whatever you want. Ex: '.menu-button', '#button'.
			opener: undefined, //STRING
			//*REQUIRED* WRAPPER is the selector for the wrapper/container div/section that holds all content. It is required to animate correctly the document. Fixed elements ("position: fixed") can be included or excluded. It does not matter if you put inside the $sidebar and the navigation-bar nor if you put them outside. Ex: '.wrapper', '#container' etc.
			wrapper: undefined, //STRING
			//IGNORE is the selector for all elements that must be ignored. This feature comes in help especially when you have to install this plug-in in a CMS or blog. TUMBLR Ex: '#ga_target, #tumblr_controls'.
			ignore: undefined, //STRING
			//DATA is the data-attribute for the new elements that the plug-in creates.
			data: 'ssbplugin',
			//ANIMATION properties.
			animation: {
				//DURATION of the animation
				duration: 500,
				//EASING of the animation. If you include the "jQuery-UI" library in your document you can specify an easing (see https://jqueryui.com/resources/demos/effect/easing.html) Ex: 'linear', 'swing', 'easeInQuad', 'easeOutQuad', 'easeInOutQuad', 'easeInCubic', 'easeOutCubic', 'easeInOutCubic', 'easeInQuart', 'easeOutQuart', 'easeInOutQuart'.
				easing: 'swing'
			}
		},
		//*REQUIRED* all SIDEBAR settings
		sidebar: {
			//*REQUIRED* the ALIGN of the $sidebar. Ex: undefined, 'left', 'right'.
			//*WARNING* type in lowercase. If undefined the align is automatically assigned to 'left'.
			align: undefined, //STRING
			//*REQUIRED* the WIDTH and MAXWIDTH of the $sidebar. This value will override CSS.
			width: 350, //NUMBER
			//the GAP between the end of the sidebar (margin-right for the left-sidebar. margin-left for the right-sidebar) and the end of the document (document margin-right for the left-sidebar. document margin-left for the right-sidebar) when the screen is smaller that the $sidebar width.
			gap: 64, //NUMBER
			//CLOSINGLINKS is the selector that will fire the $sidebar closing animation.
			//*WARNING* do not select parent and child. Ex: 'li, a' when "<li><a href=""></a></li>". So just select the child element ('a'). POSSIBLE EXAMPLES: 'a', 'li', 'a, .fauxlinks, #social, .button'.
			closingLinks: 'a',
			//*REQUIRED* This feature assures the plug-in that some values will override CSS.
			style: {
				//*REQUIRED* zINDEX or 'z-index' is required to position the sidebar above ALL content and to position the "mask div" one level below the sidebar.
				zIndex: 3000
				//
				//*WARNING* You can add more style here or in a separate file but everything here will override styles in separate CSS files.
				//*WARNING* Do no care about "position: fixed, top:0, left:0, bottom:0, right:0" because the plug-in will care about it. Positioning rules will be overridden automatically by the plug-in.
			}
		},
		//MASK holds all settings of the mask-div. The mask-div is a required feature that lays over the document content (except the $sidebar) and fires the closing animation function.
		mask: {
			//STYLE holds all CSS rules. Use this feature to stylize the mask.
			style: {
				//Default options.
				backgroundColor: 'black', //if you do not want any color use 'transparent'.
				opacity: 0.5, //if you do not want any opacity use 0.
				filter: 'Alpha(opacity=50)' //IE8 and earlier - If you do not want any opacity use 0.
				//You can add more options.
			}
		}
	});


CHANGELOG.md
CHANGELOG
============

##[v1.1.0](https://github.com/dcdeiv/simple-sidebar/releases/tag/v1.1.0)
* FIXED:
 * maskDiv double-click bug. 

##[v1.0.3](https://github.com/dcdeiv/simple-sidebar/releases/tag/v1.0.3)
* FIXED:
 * Issue [#3](https://github.com/dcdeiv/simple-sidebar/issues/3).

##[v1.0.2](https://github.com/dcdeiv/simple-sidebar/releases/tag/v1.0.2)
* FIXED:
 * Mask-div: now it is appended and not prepended to body;
 * anchor links: now they work properly;
 * right-sidebar: now it works properly;
* DELETED:
 * colsole.log;
* ADDED:
 * support for elements positioned absolute.

##[v1.0.1](https://github.com/dcdeiv/simple-sidebar/releases/tag/v1.0.1)
* FIXED: maskDiv.


jquery.simplesidebar.js
//Simple Sidebar v1.1.0 by DcDeiv https://github.com/dcdeiv
// GPLv2 http://www.gnu.org/licenses/gpl-2.0-standalone.html
(function( $ ) {
	$.fn.simpleSidebar = function( options ) {
		//declaring all global variables
		var sbw, animationStart, animationReset, callbackA, callbackB,
			//allowing user customisation
			defaults  = {
				settings: {
					opener: undefined,
					wrapper: undefined, //HTML tag is not safe. Please, use a container/wrapper div
					ignore: undefined,
					data: 'ssbplugin',
					animation: {
						duration: 500, //milliseconds (0.5s = 500ms)
						easing: 'swing'
					}
				},
				sidebar: {
					align: undefined,
					width: 350, //pixels
					gap: 64, //pixels
					closingLinks: 'a',
					style: {
						zIndex: 3000
					}
				},
				mask: {
					style: {
						backgroundColor: 'black',
						opacity: 0.5,
						filter: 'Alpha(opacity=50)' //IE8 and earlier
					}
				}
			},
			config    = $.extend( true, defaults, options ),
			$sidebar  = this,
			$opener   = $( config.settings.opener ),
			$wrapper  = $( config.settings.wrapper ),
			$ignore   = $( config.settings.ignore ),
			dataName  = config.settings.data,
			duration  = config.settings.animation.duration,
			easing    = config.settings.animation.easing,
			defAlign  = config.sidebar.align,
			sbMaxW    = config.sidebar.width,
			gap       = config.sidebar.gap,
			$links    = config.sidebar.closingLinks,
			defStyle  = config.sidebar.style,
			maskDef   = config.mask.style,
			winMaxW   = sbMaxW + gap,
			//selecting all fixed elements except the sidebar and the ignore elements
			$fixedEl  = $( '*' )
				.not( $ignore )
				.not( $sidebar )
				.filter(function() {
					return $( this ).css( 'position' ) == 'fixed';
				}),
			$absolEl  = $( '*' )
				.not( $ignore )
				.filter(function() {
					return $( this ).css( 'position' ) == 'absolute';
				}),
			//selecting all elements.
			$elements = $fixedEl
				.add( $absolEl )
				.add( $sidebar )
				.add( $wrapper )
				.not( $ignore ),
			w         = $( window ).width(),
			MaskDef   = {
				position: 'fixed',
				top: -200,
				right: -200,
				left: -200,
				bottom: -200,
				zIndex: config.sidebar.style.zIndex - 1
			},
			maskStyle = $.extend( {},  maskDef, MaskDef ),
			clicks    = 0,
			//hiding overflow [callback(A/B)]
			overflowTrue = function() {
				$( 'body, html' ).css({
					overflow: 'hidden'
				});
				
				clicks = 0;
			},
			//adding overflow [callback(A/B)]
			overflowFalse = function() {
				$( 'body, html' ).css({
					overflow: 'auto'
				});
				
				clicks = 1;
			};
		
		//adding default style to $sidebar
		$sidebar
			.css( defStyle )
			//wrapping inner content to let it overflow
			.wrapInner( '<div data-' + dataName + '="sub-wrapper"></div>' );
			
		var subWrapper = $sidebar.children().filter(function() {
			return $( this ).data( dataName ) === 'sub-wrapper' ;
		});
		
		subWrapper.css({
			width: '100%',
			height: '100%',
			overflow: 'auto'
		});
			
		//Appending to 'body' the mask-div and adding its style
		$( 'body' ).append( '<div data-' + dataName + '="mask"></div>' );
		
		var maskDiv = $( 'body' ).children().filter(function(){
			return $( this ).data( dataName ) === 'mask' ;
		});
		
		maskDiv
			.css( maskStyle )
			.hide();
		
		//assigning value to sbw
		if ( w < winMaxW ) {
			sbw = w - gap;
		} else {
			sbw = sbMaxW;
		}
		
		//testing config.sidebar.align
		if( defAlign === undefined || defAlign === 'left' ) {	
			$sidebar.css({
				position: 'fixed',
				top: 0,
				bottom: 0,
				left: 0,
				width: sbw,
				marginLeft: -sbw
			});
		} else {
			$sidebar.css({
				position: 'fixed',
				top: 0,
				right: 0,
				bottom: 0,
				width: sbw,
				marginRight: -sbw
			});
		}
		
		$opener.click(function() {
			var nsbw = $sidebar.width();
			
			if( defAlign === undefined || defAlign === 'left' ) {
				animationStart = {
					marginLeft: '+=' + nsbw,
					marginRight: '-=' + nsbw
				};
			} else {
				animationStart = {
					marginRight: '+=' + nsbw,
					marginLeft: '-=' + nsbw
				};
			}
			
			$elements.each(function() {
				$( this ).animate( animationStart, {
					duration: duration,
					easing: easing,
					complete: overflowTrue
				});
			});
			
			maskDiv.fadeIn();
		});
		
		maskDiv.click(function() {
			clicks++;
			var nsbw = $sidebar.width();
				countClicks = function( e ) {
					return ( e % 2 === 0 ) ? true : false;
				};
			
			if( defAlign === undefined || defAlign === 'left' ) {
				animationStart = {
					marginLeft: '+=' + nsbw,
					marginRight: '-=' + nsbw
				};
				animationReset = {
					marginRight: '+=' + nsbw,
					marginLeft: '-=' + nsbw
				};
			} else {
				animationStart = {
					marginRight: '+=' + nsbw,
					marginLeft: '-=' + nsbw
				};
				animationReset = {
					marginLeft: '+=' + nsbw,
					marginRight: '-=' + nsbw
				};
			}
			
			if ( false === countClicks ( clicks ) ) {
				$elements.each(function() {
					$( this ).animate( animationReset, {
						duration: duration,
						easing: easing,
						complete: overflowFalse
					});
				});
				
				maskDiv.fadeOut();
			} else if ( true === countClicks ( clicks ) ) {
				$elements.each(function() {
					$( this ).animate( animationStart, {
						duration: duration,
						easing: easing,
						complete: overflowTrue
					});
				});
				
				maskDiv.fadeIn();
			}
		});
		
		$sidebar.on( 'click', $links, function() {
			var nsbw = $sidebar.width();
			
			if( defAlign === undefined || defAlign === 'left' ) {
				animationReset = {
					marginRight: '+=' + nsbw,
					marginLeft: '-=' + nsbw
				};
			} else {
				animationReset = {
					marginLeft: '+=' + nsbw,
					marginRight: '-=' + nsbw
				};
			}
			
			$elements.each(function() {
				$( this ).animate( animationReset, {
					duration: duration,
					easing: easing,
					complete: overflowFalse
				});
				
				maskDiv.fadeOut();
			});
			
		});
		
		//Adding responsive to $sidebar
		$( window ).resize(function() {
			var rsbw, sbMar,
				w = $( this ).width();
				
			if ( w < winMaxW ) {
				rsbw = w - gap;
			} else {
				rsbw = sbMaxW;
			}
			
			$sidebar.css({
				width: rsbw
			});
			
			//fixing $element position according to $sidebar new width (rsbw)
			if ( defAlign === undefined || defAlign === 'left' ) {
				sbMar = parseInt( $sidebar.css( 'margin-left' ) );
				
				if ( 0 > sbMar ) {
					$sidebar.css({
						marginLeft: -rsbw
					});
				} else {
					$elements.not( $sidebar )
						.css({
							marginLeft: + rsbw,
							marginRight: - rsbw
						});
				}
			} else {
				sbMar = parseInt( $sidebar.css( 'margin-right' ) );
				
				if ( 0 > sbMar ) {
					$sidebar.css({
						marginRight: -rsbw
					});
				} else {
					$elements.not( $sidebar )
						.css({
							marginLeft: - rsbw,
							marginRight: + rsbw
						});
				}
			}
		});
		
		return this;
	};
})( jQuery );
